symbulate/results.py

```diff
@@ -330,61 +365,216 @@ def plot(self, type=None, alpha=None, normalize=True, jitter=False,
+                cbar.set_label('Relative Frequency')
+                caxes.yaxis.set_label_position("left")
+            elif 'mixed-tile' in type:
+                if not discrete_x:
+                    x_bin = np.linspace(min(x), max(x) + 1, bins)
+                    x_lab = [min(x)]
+                    for i in range(len(x_bin) - 1):
+                        x_lab.append(0.5 * (x_bin[i] + x_bin[i+1]))
+                    x_pos = range(len(x_lab))
+                    x = np.digitize(x, x_bin)
+                else:
+                    x_lab = np.unique(x)
+                    x_pos = range(len(x_lab))
+                    x_map = dict(zip(x_lab, x_pos))
+                    x = np.vectorize(x_map.get)(x)
+
+                if not discrete_y:
+                    y_bin = np.linspace(min(y), max(y) + 1, bins)
+                    y_lab = [min(y)]
+                    for i in range(len(y_bin) - 1):
+                        y_lab.append(0.5 * (y_bin[i] + y_bin[i+1]))
+                    y = np.digitize(y, y_bin)
+                    y_pos = range(len(y_lab))
+                else:
+                    y_lab = np.unique(y)
+                    y_pos = range(len(y_lab))
+                    y_map = dict(zip(y_lab, y_pos))
+                    y = np.vectorize(y_map.get)(y)
+
+                nums = len(x)
+
+                counts = count_var(list(zip(y, x)))
+                intensity = np.zeros(shape=(len(y_lab), len(x_lab)))
+                    
+                for key, val in counts.items():
+                    intensity[key] = val / nums
+                hm = ax.matshow(intensity, cmap='Blues', origin='lower', aspect='auto')
+                ax.xaxis.set_ticks_position('bottom')
+                ax.set_xticks(x_pos)
+                ax.set_yticks(y_pos)
+                if not discrete_x: x_lab = np.around(x_lab, decimals=1)
+                if not discrete_y: y_lab = np.around(y_lab, decimals=1)
+                ax.set_xticklabels(x_lab)
+                ax.set_yticklabels(y_lab)
+
+                if 'marginal' not in type:
+                    caxes = fig.add_axes([0, 0.1, 0.05, 0.8])
+                else:
+                    caxes = fig.add_axes([0, 0.1, 0.05, 0.57])
+                cbar = plt.colorbar(mappable=hm, cax=caxes)
+                caxes.yaxis.set_ticks_position('left')
+                cbar.set_label('Relative Frequency')
+                caxes.yaxis.set_label_position("left")
+            elif 'violin' in type:
+                res = np.array(self)
+                values = []
+                if discrete_x and not discrete_y:
+                    positions = sorted(list(x_count.keys()))
+                    for i in positions:
+                        values.append(res[res[:, 0] == i, 1].tolist())
+                    violins = ax.violinplot(dataset=values, showmedians=True)
+                    ax.set_xticks(np.array(positions) + 1)
+                    ax.set_xticklabels(positions)
+                else: 
```

Comment:
Do you see how the code in this `else` block is basically the same as the above code, except with "x" replaced by "y"? This is a good candidate for code to factor into plot.py. You can write a function `make_violin_plot()` that takes in the data and which axis ("x" or "y") is discrete and then generates the appropriate plot.

I encourage you to also write a helper function `set_ticks()` in plot.py that takes in the positions, labels, and which axis ("x" or "y").